### PR TITLE
[ETL-505] Add Garmin transforms to S3 to JSON job

### PR DESCRIPTION
**Pull request by philerooski**
In addition to the Garmin transforms, I refactored the S3 to JSON job to be more functional. What used to just be
* `write_file_to_json_dataset`

has been split up into three different functions:
* `transform_json` - which takes care of any inserts/updates to the JSON data
* `transform_object_to_array_of_objects` - which is a helper function for `transform_json`
* `get_output_filename` - which takes care of filename / S3 object name formatting.

As a result of refactoring in the job, the tests have been refactored as well. Some tests were deleted and replaced with a more specific test.

### CODE REVIEW

File: src/glue/jobs/s3_to_json.py [hunk 1]
```diff
@@ -16,6 +16,237 @@
...
+        # Resulting `transformed_array`:
+        # [
+        #     {'OffsetInSeconds': 0, 'HeartRate': 62},
+        #     {'OffsetInSeconds': 1, 'HeartRate': 63},
+        #     {'OffsetInSeconds': 2, 'HeartRate': 62},
+        # ]
+    """
+    array_of_obj = [
+            {
+                key_name: key_type(k),
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:74]: Will there ever be cases where you won't be able to cast the value?

Things like (int('')) might cause an error.

File: src/glue/jobs/s3_to_json.py [hunk 2]
```diff
@@ -16,6 +16,237 @@
...
+
+    Args:
+        json_obj (str): A JSON object sourced from the JSON file of this data type.
+        dataset_identifier (str): The data type of `json_obj`.
+        metadata (dict): Metadata derived from the file basename.
+
+    Returns:
+        json_obj (dict) The JSON object with the relevant transformations applied.
+    """
+    logger = logging.getLogger(__name__)
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:113]: Why not set the logger globally?

File: src/glue/jobs/s3_to_json.py [hunk 3]
```diff
@@ -16,6 +16,237 @@
...
+    if metadata["start_date"] is not None:
+        json_obj["export_start_date"] = metadata["start_date"].isoformat()
+    else:
+        json_obj["export_start_date"] = None
+    json_obj["export_end_date"] = metadata.get("end_date").isoformat()
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:118]: Is there always an end date? Does .isoformat() work when the value is None?

File: src/glue/jobs/s3_to_json.py [hunk 4]
```diff
@@ -16,6 +16,237 @@
...
+    if dataset_identifier in DATA_TYPES_WITH_SUBTYPE:
+        # This puts the `Type` property back where Apple intended it to be
+        json_obj["Type"] = metadata["subtype"]
+    if dataset_identifier == "SymptomLog":
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:122]: Can these be else ifs?

File: src/glue/jobs/s3_to_json.py [hunk 5]
```diff
@@ -16,6 +16,237 @@
...
+        # Load JSON string as dict
+        json_obj["Value"] = json.loads(json_obj["Value"])
+    if dataset_identifier == "EnrolledParticipants":
+        for field_name in ["Symptoms", "Treatments"]:
+            if (
+                    field_name in json_obj["CustomFields"]
+                    and isinstance(json_obj["CustomFields"][field_name], str)
+               ):
+                if len(json_obj["CustomFields"][field_name]) > 0:
+                    # This JSON string was written in a couple different ways
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:132]: Wow, this is nasty. Is this worth letting care evolution know about this?

File: src/glue/jobs/s3_to_json.py [hunk 6]
```diff
@@ -16,6 +16,237 @@
...
+                    except json.JSONDecodeError as e:
+                        # If it's not propertly formatted JSON, then we
+                        # can't read it, and instead store an empty list
+                        logger.warning(f"Problem CustomFields.{field_name}: "
+                                       f"{json_obj['CustomFields'][field_name]}")
+                        logger.warning(str(e))
+                        json_obj["CustomFields"][field_name] = []
+                else:
+                    json_obj["CustomFields"][field_name] = []
+    garmin_transform_types = {
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:150]: I like what you've done here, but is it higher risk in maintenance down the line if we forget that the first tuple has to be the key and the second tuple is value?

File: src/glue/jobs/s3_to_json.py [hunk 7]
```diff
@@ -16,6 +16,237 @@
...
+                if prop_name in json_obj:
+                    array_of_obj = transform_object_to_array_of_objects(
+                            json_obj_to_replace=json_obj[prop_name],
+                            key_name=key_name,
+                            key_type=key_type,
+                            value_name=value_name,
+                            value_type=value_type
+                    )
+                    json_obj[prop_name] = array_of_obj
+            if len(property_hierarchy) == 2:
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:197]: Nit: the logic under this if statement sort of resembles the logic above : can they be combined to simplify the code?

File: src/glue/jobs/s3_to_json.py [hunk 8]
```diff
@@ -16,6 +16,237 @@
...
+        - start_date
+        - end_date
+
+    Args:
+        metadata (dict): Metadata derived from the file basename.
+
+    Return:
+        str: A formatted file name.
+    """
+    if metadata["type"] in DATA_TYPES_WITH_SUBTYPE:
```
Comment (thomasyu888) [inline src/glue/jobs/s3_to_json.py:230]: I know this was code that already existed but does datatypes with subtypes always have startdate?

File: tests/test_s3_to_json.py [hunk 9]
```diff
@@ -83,117 +83,309 @@ def json_file_basenames_dict(self):
...
+                "end_date": datetime.datetime(2023, 1, 14, 0, 0),
+        }
+        transformed_json = s3_to_json.transform_json(
+                json_obj={},
+                dataset_identifier=sample_metadata["type"],
+                metadata=sample_metadata
+        )
+
+        assert (
+            sample_metadata["start_date"].isoformat()
```
Comment (thomasyu888) [inline tests/test_s3_to_json.py:128]: You may want to explicitly write out the isoformat(). This is because if ever (isoformat) has a bug, this test would pass.  (Highly unlikely but just something to think about when writing tests that use the same functions as the code you're testing.

File: tests/test_s3_to_json.py [hunk 10]
```diff
@@ -83,117 +83,309 @@ def json_file_basenames_dict(self):
...
+        transformed_json = s3_to_json.transform_json(
+                json_obj={
+                    "Summaries": [
+                        {
+                            "EpochSummaries": epoch_summaries,
+                            "Dummy": 1
+                        },
+                        {
+                            "EpochSummaries": epoch_summaries,
+                            "Dummy": 1
```
Comment (thomasyu888) [inline tests/test_s3_to_json.py:320]: Should this dummy value be different?

File: tests/test_s3_to_json.py [hunk 11]
```diff
@@ -83,117 +83,309 @@ def json_file_basenames_dict(self):
...
-        with zipfile.ZipFile(io.BytesIO(s3_obj["Body"])) as z:
-            output_file = s3_to_json.write_file_to_json_dataset(
-                z=z,
-                json_path="SymptomLog_20230112-20230114.json",
-                dataset_identifier="SymptomLog",
-                metadata=sample_metadata["Metadata"],
-                workflow_run_properties=workflow_run_properties,
-                delete_upon_successful_upload=False,
-            )
+        transformed_epoch_summaries = [
```
Comment (thomasyu888) [inline tests/test_s3_to_json.py:292]: This is the expected transform right?

File: src/glue/jobs/s3_to_json.py (hunk 1 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:74]: I don't know. This data comes from Garmin (I believe CareEvolution is just an intermediary, they don't touch this part of the data), so it depends on what processes Garmin has in place so that they aren't writing incorrect types. I didn't experience any issues with the pilot data.

File: src/glue/jobs/s3_to_json.py (hunk 3 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:118]: The end date is derived from the filename. It's either included (in bold) like GarminHealthSnapshotSummary_20230626-**20230627**.json or EnrolledParticipants_**20230627**.json

Care Evolution always formats their exports like this. So, yes, there is always an end date.

File: src/glue/jobs/s3_to_json.py (hunk 4 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:122]: Either this way or with else if's would work. The conditions are predicated on the data type and each condition is mutually exclusive. Potentially in the future we will add non-mutually exclusive conditions (like `if dataset_identifier == data_type` and then later in the code... `if property == value`), in which case we wouldn't want to use else if, or only one of the conditional blocks will be executed.

File: src/glue/jobs/s3_to_json.py (hunk 5 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:132]: I haven't seen the `[{\\\"id\\\": ...` format in the production data (we would log a warning if we ran into a `JSONDecodeError`), so whatever issues they had with the pilot data seem to have been resolved.

File: src/glue/jobs/s3_to_json.py (hunk 7 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:197]: This was something I did intentionally. It seems pretty difficult to write this in a way that can handle an arbitrary amount of nesting of the JSON property. I would need to write a recursive function. The `len == 1` case would be the base case and you can begin to imagine how the recursive case might look with the extra `for...` block in the `len == 2` case. I think a recursive implementation would be more difficult to understand/debug, so unless we need to handle deeper nesting in the future I think it's better to have it implemented more explicitly -- although a bit redundantly.

File: src/glue/jobs/s3_to_json.py (hunk 8 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py:230]: Yes, so far.

File: tests/test_s3_to_json.py (hunk 9 above)
Comment (philerooski) [reply tests/test_s3_to_json.py:128]: That's an interesting edge case that `isoformat` might have a bug, although it's from the python standard library. I don't think it's an edge case we need to test for. That would be up to python's developers to test for bugs in `datetime`. 

Perhaps a more likely scenario is that the default ISO format outputted by `isoformat` changes. I would be okay with that, as long as it's ISO standard.

File: tests/test_s3_to_json.py (hunk 10 above)
Comment (philerooski) [reply tests/test_s3_to_json.py:320]: That's not really necessary, but I've removed `Dummy` from the second object anyways since we don't use it as part of an assert statement.

File: tests/test_s3_to_json.py (hunk 11 above)
Comment (philerooski) [reply tests/test_s3_to_json.py:292]: Yes.

File: src/glue/jobs/s3_to_json.py (hunk 2 above)
Comment (philerooski) [reply src/glue/jobs/s3_to_json.py]: That's fair. I've made it global now.

File: src/glue/jobs/s3_to_json.py (hunk 4 above)
Comment (thomasyu888) [reply src/glue/jobs/s3_to_json.py:124]: That makes sense.  This leaves room open for future conditions.

File: src/glue/jobs/s3_to_json.py (hunk 7 above)
Comment (thomasyu888) [reply src/glue/jobs/s3_to_json.py:199]: That makes sense.  Can you add a comment here about that?

File: tests/test_s3_to_json.py (hunk 9 above)
Comment (thomasyu888) [reply tests/test_s3_to_json.py:128]: Makes sense to me!

File: src/glue/jobs/s3_to_json.py (hunk 5 above)
Comment (thomasyu888) [reply src/glue/jobs/s3_to_json.py:134]: We just have to make sure we are monitoring the logs for these warnings.

**Review by thomasyu888 — APPROVED** [review]
💯 awesome!
